torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py

```diff
@@ -0,0 +1,818 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" PyTorch GPTNeoX model."""
+
+from typing import Optional, Tuple, Union
+
+import torch
+import torch.utils.checkpoint
+from torch import nn
+
+from transformers.activations import ACT2FN
+from transformers.file_utils import (
+    add_code_sample_docstrings,
+    add_start_docstrings,
+    add_start_docstrings_to_model_forward,
+    replace_return_docstrings,
+)
+from transformers.modeling_outputs import BaseModelOutputWithPast, CausalLMOutputWithPast
+from transformers.modeling_utils import PreTrainedModel
+from transformers.utils import logging
+from transformers.models.gpt_neox.configuration_gpt_neox import GPTNeoXConfig
+
+from neuronx_distributed.parallel_layers.layers import ParallelEmbedding, ColumnParallelLinear, RowParallelLinear
+from neuronx_distributed.parallel_layers.loss_functions import parallel_cross_entropy
+from neuronx_distributed.parallel_layers.parallel_state import get_tensor_model_parallel_size
+import neuronx_distributed.parallel_layers.utils as neuronx_dist_utils
+from neuronx_distributed.parallel_layers import move_model_to_device, mappings, layer_norm
+import torch_xla.core.xla_model as xm
+
+from functools import partial
+def _init_normal(std, w):
+    return nn.init.normal_(w, mean=0.0, std=std)
+
+logger = logging.get_logger(__name__)
+
+_CHECKPOINT_FOR_DOC = "trl-internal-testing/tiny-random-GPTNeoXForCausalLM"
+_REAL_CHECKPOINT_FOR_DOC = "EleutherAI/gpt-neox-20b"
+_CONFIG_FOR_DOC = "GPTNeoXConfig"
+
+GPT_NEOX_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "EleutherAI/gpt-neox-20b",
+    # See all GPTNeoX models at https://huggingface.co/models?filter=gpt_neox
+]
+
+
+class GPTNeoXPreTrainedModel(PreTrainedModel):
+    """
+    An abstract class to handle weights initialization and a simple interface for downloading and loading pretrained
+    models.
+    """
+
+    config_class = GPTNeoXConfig
+    base_model_prefix = "gpt_neox"
+    supports_gradient_checkpointing = True
+    _no_split_modules = ["GPTNeoXLayer"]
+
+    def _init_weights(self, module):
+        """Initialize the weights"""
+        if isinstance(module, nn.Linear):
+            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
+            if module.bias is not None:
+                module.bias.data.zero_()
+        elif isinstance(module, nn.Embedding):
+            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
+            if module.padding_idx is not None:
+                module.weight.data[module.padding_idx].zero_()
+        elif isinstance(module, nn.LayerNorm):
+            module.bias.data.zero_()
+            module.weight.data.fill_(1.0)
+
+    def _set_gradient_checkpointing(self, module, value=False):
+        if isinstance(module, GPTNeoXModel):
+            module.gradient_checkpointing = value
+
+
+class GPTNeoXAttention(nn.Module):
```

Comment:
Are we touching each an every function of the original class. Just checking if we can inherit the original class and override those functions which we need. This way we can show the users what are the main changes and where we deviate. We can add comments too in the places we add new changes

Reply:
Have cleaned up the code by deleting unnecessary codes.